httpbp/client_middlewares.go

```diff
@@ -0,0 +1,194 @@
+					return nil, err
+				}
+				// circuit break on any HTTP 5xx code
+				if resp.StatusCode >= http.StatusInternalServerError {
+					err = resp.Body.Close()
+					if err != nil {
+						return nil, err
+					}
+					return nil, err
+				}
+				return resp, nil
+			})
+			if err != nil {
+				return nil, err
+			}
+			return resp.(*http.Response), nil
+		})
+	}
+}
+
+func applyMiddlewares(transport http.RoundTripper, middlewares ...ClientMiddleware) http.RoundTripper {
+	// add middlewares in reverse so the first in the list is the outermost
+	for i := len(middlewares) - 1; i >= 0; i-- {
+		transport = middlewares[i](transport)
+	}
+	return transport
+}
+
+// Retries provides a retry middleware.
+func Retries(config retry.Config) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+			retrybp.Do(req.Context(), func() error {
+				resp, err = next.RoundTrip(req)
+				if err != nil {
+					return err
+				}
+				return nil
+			}, retry.Attempts(1))
+			if err != nil {
+				return nil, err
+			}
+			return resp, nil
+		})
+	}
+}
+
+// LoadBalancer implements a round-robin load balancer.
+func LoadBalancer(config Config) ClientMiddleware {
+	attempts := 0
+	offset := uint64(0)
+	hosts := config.LoadBalancer.Hosts
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+
+			remaining := attempts
+			for remaining > 0 {
+				i := atomic.AddUint64(&offset, 1)
+				hostIndex := i % uint64(len(hosts))
+				req.Host = hosts[hostIndex]
+				req.URL.Host = hosts[hostIndex]
+
+				resp, err := next.RoundTrip(req)
+				if err == gobreaker.ErrOpenState || err == gobreaker.ErrTooManyRequests {
+					remaining--
+				} else {
+					return resp, err
+				}
+			}
+			return next.RoundTrip(req)
+		})
+	}
+}
+
+// MaxConcurrency is a middleware to ensure that there will be only a maximum
+// number of requests concurrently in-flight at any given time.
+func MaxConcurrency(config Config) ClientMiddleware {
+	maxConcurrency := int64(config.MaxConcurrency)
+	activeRequests := int64(0)
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+			attemptedRequests := atomic.AddInt64(&activeRequests, 1)
+			defer atomic.AddInt64(&activeRequests, -1)
+
+			if maxConcurrency > 0 && attemptedRequests > maxConcurrency {
+				return nil, ErrConcurrencyLimit
+			}
+			return next.RoundTrip(req)
+		})
+	}
+}
+
+// MonitorClient is a HTTP client middleware that wraps HTTP requests in a
+// client span.
+func MonitorClient(config Config) ClientMiddleware {
+	slug := config.Slug
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+			span, ctx := opentracing.StartSpanFromContext(
+				req.Context(),
+				slug,
```

Comment:
for client spans, we do expect the name to be in the format of `client.endpoint`.

since for http requests the "endpoint" part is often ambiguous, I'd suggest either:

1. hardcode an endpoint for everything, e.g. `slug + ".http"`

or

2. convert the uri to endpoint, e.g. `slug + "." + sanitize(req.URL.Path)` (`sanitize` should replace `/`, `.`, and a few others to `_` or `-`).

Reply:
+1 on that but I think this is something we should coordinate with baseplate.py? https://github.com/reddit/baseplate.py/blob/48bcb0b099608aa8b8eecbf97cb929537d68753d/baseplate/clients/requests.py#L203

Reply:
that's basically my proposal 1, but with `request` instead of `http` as the hardcoded endpoint name :)

Reply:
```suggestion
				slug + ".request",
```